six/modules/python/cphd/source/cphd.i

```diff
@@ -193,26 +511,134 @@ using six::Vector3;
 
 %extend cphd::CPHDWriter
 {
+
 %pythoncode
 %{
     def __del__(self):
         self.close()
 %}
+
+/*! Write a sequence of NumPy arrays of wideband data to a CPHD file
+ * \param metadata (cphd::Metadata)
+ * \param pvpBlock (cphd::PVPBlock)
+ * \param widebandArray
+ *        A sequence of NumPy arrays of wideband data, one array per channel
+ * \param supportData (sys::ubyte*)
+          Defaults to nullptr
+ */
+void writeWideband(const Metadata& metadata,
+                   const PVPBlock& pvpBlock,
+                   PyObject* widebandArrays,
+                   const sys::ubyte* supportData=nullptr)
+{
+    const size_t numChannels = static_cast<size_t>(PySequence_Length(widebandArrays));
+    if (numChannels != metadata.data.getNumChannels())
+    {
+        std::ostringstream oss;
+        oss << "Number of channels in metadata (" << metadata.data.getNumChannels() << ") "
+            << "does not match number of channels received (" << numChannels << ")";
+        throw except::Exception(Ctxt(oss.str()));
+    }
+
+    $self->writeMetadata(pvpBlock);
+
+    if (metadata.data.getNumSupportArrays() != 0)
+    {
+        if (supportData == nullptr)
+        {
+            throw except::Exception(Ctxt("SupportData is not provided"));
+        }
+        $self->writeSupportData(supportData);
+    }
+
+    $self->writePVPData(pvpBlock);
+
+    for (size_t ii = 0; ii < numChannels; ++ii)
+    {
+        PyObject* channelWideband = PySequence_GetItem(widebandArrays, ii);
+
+        // Note that different channels could have different dimensions
+        const types::RowCol<size_t> metadataDims(
+            metadata.data.getNumVectors(ii),
+            metadata.data.getNumSamples(ii)
+        );
+        const types::RowCol<size_t> arrayDims = numpyutils::getDimensionsRC(channelWideband);
+        if (metadataDims != arrayDims)
+        {
+            std::ostringstream oss;
+            oss << "For channel " << ii << ", wideband dimensions ("
+                << arrayDims.row << ", " << arrayDims.col << ") "
+                << "do not match dimensions in metadata: ("
+                << metadataDims.row << ", " << metadataDims.col << ")";
+            throw except::Exception(Ctxt(oss.str()));
+        }
+
+        const size_t numElements = arrayDims.row * arrayDims.col;
+        std::complex<float>* widebandPtr = numpyutils::getBuffer<std::complex<float>>(channelWideband);
+        $self->writeCPHDData<std::complex<float>>(widebandPtr, numElements, ii);
+    }
+}
+
 }
 
+%pythoncode
+%{
+    import os
+    import numpy as np
+
+    def write_cphd(
+        pathname,
+        metadata,
+        pvp_block,
+        wideband_arrays,
+        schema_paths=[],
```

Comment:
Avoid mutable default arguments (i.e. lists, dicts, etc.). Python doesn't make a copy when the function is invoked as C++ would. Some squirrely behavior can result. Instead use `None` and parse what `None` means inside the function body.